### PR TITLE
chore(neurons_fund): Populate hotkeys when necessary in the NNS Governance → Swap → SNS Governance dataflow

### PR DESCRIPTION
**Pull request by anchpop**
Neurons' Fund neuron information starts in NNS Governance, goes to Swap, then goes to SNS Governance. There were some places where we did not populate the hotkeys field, along this chain. This PR rectifies that, although it doesn't yet actually "complete" the chain as that is done in #629

### CODE REVIEW

File: rs/nns/governance/src/governance.rs [hunk 1]
```diff
@@ -289,7 +289,7 @@ impl From<NeuronsFundNeuronPortion> for NeuronsFundNeuronPortionPb {
             maturity_equivalent_icp_e8s: Some(neuron.maturity_equivalent_icp_e8s),
             controller: Some(neuron.controller),
             is_capped: Some(neuron.is_capped),
-            hotkeys: Vec::new(), // TODO(NNS1-3199): populate hotkeys
+            hotkeys: neuron.hotkeys,
```
Comment (anchpop) [inline rs/nns/governance/src/governance.rs:292]: TODO: add test

File: rs/sns/swap/src/swap.rs [hunk 2]
```diff
@@ -911,8 +912,7 @@ impl Swap {
 
                     match create_sns_neuron_basket_for_cf_participant(
                         &controller,
-                        // TODO(NNS1-3199): Populate this field
-                        Vec::new(),
+                        hotkeys.principals,
```
Comment (anchpop) [inline rs/sns/swap/src/swap.rs:915]: TODO: add a test

File: rs/nervous_system/common/src/lib.rs [hunk 3]
```diff
@@ -83,6 +83,15 @@ const WASM_PAGE_SIZE_BYTES: usize = 65536;
 // 1 Mi. Approximately 10^6, 1 million (slightly more).
 const MAX_LOGS_RESPONSE_SIZE: usize = 1 << 20;
 
+/// Generates as many user test ids as requested, starting from `starting_index`.
+/// ```rust
+/// // generate 3 test IDs, with indices starting at 55:
+/// let [id1, id2, id3] = new_user_test_ids(55);
+/// ```
+pub fn new_user_test_ids<const N: usize>(starting_index: u64) -> [PrincipalId; N] {
```
Comment (aterga) [inline rs/nervous_system/common/src/lib.rs:91]: Is this supposed to be a test-only function? If so, let's not declare it in this crate. At the very least, test functions should be compiled only for tests.

File: rs/sns/swap/src/swap.rs [hunk 4]
```diff
@@ -881,7 +881,8 @@ impl Swap {
 
             for cf_neuron in cf_participant.cf_neurons.iter_mut() {
                 // Create a closure to ensure `global_cf_memo` is incremented in all cases
-                let mut process_cf_neuron = || {
+                let hotkeys = cf_neuron.hotkeys.clone().unwrap_or_default();
```
Comment (aterga) [inline rs/sns/swap/src/swap.rs:884]: Would deconstructing the `cf_neuron` neuron instance allow avoiding `clone()` in this case?

File: rs/sns/swap/src/swap.rs [hunk 5]
```diff
@@ -881,7 +881,8 @@ impl Swap {
 
             for cf_neuron in cf_participant.cf_neurons.iter_mut() {
```
Comment (aterga) [inline rs/sns/swap/src/swap.rs:882]:
```suggestion
            for neurons_fund_neuron in neurons_fund_participant.cf_neurons.iter_mut() {
```

File: rs/sns/swap/src/swap.rs [hunk 6]
```diff
@@ -881,7 +881,8 @@ impl Swap {
...
                 // Create a closure to ensure `global_cf_memo` is incremented in all cases
-                let mut process_cf_neuron = || {
+                let hotkeys = cf_neuron.hotkeys.clone().unwrap_or_default();
+                let process_cf_neuron = || {
```
Comment (aterga) [inline rs/sns/swap/src/swap.rs:885]:
```suggestion
                let process_neurons_fund_neuron = || {
```

**Review by aterga — CHANGES_REQUESTED** [review]
LGTM modulo unit tests covering the new hotkey propagation code.

File: rs/nervous_system/common/src/lib.rs (hunk 3 above)
Comment (anchpop) [reply rs/nervous_system/common/src/lib.rs]: I removed the function

File: rs/sns/swap/src/swap.rs (hunk 5 above)
Comment (anchpop) [reply rs/sns/swap/src/swap.rs]: Done

File: rs/sns/swap/src/swap.rs (hunk 6 above)
Comment (anchpop) [reply rs/sns/swap/src/swap.rs]: Done

File: rs/sns/swap/src/swap.rs (hunk 4 above)
Comment (anchpop) [reply rs/sns/swap/src/swap.rs]: It seems to be not very easy to remove the .clone()

**Comment by andrew-lee-work** [comment]
LGTM

**Pull request merged**